modin/pandas/indexing.py

```diff
@@ -689,49 +768,82 @@ def _compute_lookup(self, row_loc, col_loc):
             List of columns labels.
-        """
-        row_loc = [row_loc] if is_scalar(row_loc) else row_loc
-        col_loc = [col_loc] if is_scalar(col_loc) else col_loc
-        if is_list_like(row_loc) and len(row_loc) == 1:
-            if (
-                isinstance(self.qc.index.values[0], np.datetime64)
-                and type(row_loc[0]) != np.datetime64
-            ):
-                row_loc = [pandas.to_datetime(row_loc[0])]
 
-        if isinstance(row_loc, slice):
-            row_lookup = self.qc.index.get_indexer_for(
-                self.qc.index.to_series().loc[row_loc]
-            )
-        elif self.qc.has_multiindex():
-            if isinstance(row_loc, pandas.MultiIndex):
-                row_lookup = self.qc.index.get_indexer_for(row_loc)
-            else:
-                row_lookup = self.qc.index.get_locs(row_loc)
-        elif is_boolean_array(row_loc):
-            # If passed in a list of booleans, we return the index of the true values
-            row_lookup = [i for i, row_val in enumerate(row_loc) if row_val]
-        else:
-            row_lookup = self.qc.index.get_indexer_for(row_loc)
-        if isinstance(col_loc, slice):
-            col_lookup = self.qc.columns.get_indexer_for(
-                self.qc.columns.to_series().loc[col_loc]
-            )
-        elif self.qc.has_multiindex(axis=1):
-            if isinstance(col_loc, pandas.MultiIndex):
-                col_lookup = self.qc.columns.get_indexer_for(col_loc)
+        Notes
+        -----
+        Usage of `slice(None)` as a resulting lookup is a hack to pass information about
+        full-axis grab without computing actual indices that triggers lazy computations.
+        Ideally, this API should get rid of using slices as indexers and either use a
+        common ``Indexer`` object or range and ``np.ndarray`` only.
+        """
+        lookups = []
+        for axis, axis_loc in enumerate((row_loc, col_loc)):
+            if is_scalar(axis_loc):
+                axis_loc = np.array([axis_loc])
+            if isinstance(axis_loc, slice) or is_range_like(axis_loc):
+                if isinstance(axis_loc, slice) and axis_loc == slice(None):
+                    axis_lookup = axis_loc
+                else:
+                    axis_labels = self.qc.get_axis(axis)
+                    # `slice_indexer` returns a fully-defined numeric slice for a non-fully-defined labels-based slice
+                    axis_lookup = axis_labels.slice_indexer(
+                        axis_loc.start, axis_loc.stop, axis_loc.step
+                    )
+                    # Converting negative indices to its actual positions:
+                    axis_lookup = pandas.RangeIndex(
+                        start=(
+                            axis_lookup.start
+                            if axis_lookup.start >= 0
+                            else axis_lookup.start + len(axis_labels)
+                        ),
```

Comment:
I wonder if doing something like
```python
start=(axis_lookup.start + len(axis_labels) % len(axis_labels),
```
is easier to read? is it faster?..

Reply:
This doesn't cover a single corner case when we're slicing with a negative step up to the first element of the index, passing such slice to the `slice_indexer` returns something like this:
```python
>>> index
Index(['a', 'b', 'c', 'd'], dtype='object')
>>> index.slice_indexer("d", None, -1)
slice(3, -5, -1)
```
The `stop` is "-5" so the conversion from negative to positive indices would return "-1" which is a proper `stop` bound for this slice ([in-code comment from pandas](https://github.com/pandas-dev/pandas/blob/c58d8ee6e93fc1ba7aa4a40dc71488cc9807ed43/pandas/core/indexes/base.py#L6389-L6395)). Since we want a range to store `[len(df) - 1, len(df) - 2, ..., 1, 0]` we have to do a `range(len(df) - 1, -1, -1)`. For the presented case, code from your suggestion will do the following:
```
1. ((- len(df) - 1) + len(df)) % len(df)
2. -1 % len(df)
3. len(df) - 1 # Not the result we wanted
```



P.S. I always thought that the actual way of implementing negative-to-positive conversion is something like `index % container_length`, can we rely on the conversion from the pandas' comment or this is a hack over a hack?

Reply:
then I wonder if we can just use that `.slice_locs()` from pandas somehow...

Reply:
[`slice_indexer`](https://github.com/pandas-dev/pandas/blob/945c9ed766a61c7d2c0a7cbb251b6edebf9cb7d5/pandas/core/indexes/base.py#L5639) is just a wrapper for `slice_locs` with some additional checks. So, actually, we're already using it.